### PR TITLE
[Fix] `preserveSymlinks: false` ensure that files are realpathed

### LINKED ISSUE #196: preserveSymlinks doesn't work properly from root package
Consider a library `lib` with the following file structure:

```
root
- common
  - node_modules
    - resolve
    - buffer
- mylib
  - node_modules
    - buffer -> ../../common/node_modules/buffer
  - src
    - index.js
```

In other words, a dependency exists from the root library to `buffer` which is symlinked into a separate folder (similar to how pnpm works).

In this case, `resolve` doesn't seem to be mapping the symlinked paths to real paths properly.

**Repro Steps**
1. Create the above example with mkdir/ln -s
2. cd into root/mylib/src, run node.exe
3. import `resolve` and observe that `resolve('buffer/')` returns the symlinked path (`root/mylib/node_modules/buffer`) instead of the real path (`root/common/node_modules/buffer`).

### PR DESCRIPTION
**Pull request by ljharb**
Fixes #196.

This very well may not have the correct behavior of `--preserve-symlinks-main`; I think we'll need some more tests before landing this.

### CODE REVIEW

**Comment by ljharb** [comment]
cc @bterlson @octogonz

File: .eslintrc [hunk 1]
```diff
@@ -27,5 +27,13 @@
         "object-curly-newline": 0,
         "operator-linebreak": [2, "before"],
         "sort-keys": 0,
-    }
+    },
+    "overrides": [
+        {
+            "files": "test/resolver/nested_symlinks/mylib/*.js",
+            "rules": {
+                "no-throw-literal": 0,
```
Comment (octogonz) [inline .eslintrc]: Instead of ` throw 'async: no match';` you could just do ` throw new Error('async: no match');`

Comment (ljharb) [reply .eslintrc]: sure, but then the failure output includes a stack trace, which makes it harder to read :-)

**Comment by octogonz** [comment]
I tested this in a PNPM repo and confirmed that `resolve.sync()` now returns the same path as `require.resolve()` for various inputs that previously were different.

**Comment by bterlson** [comment]
This PR looks good to me. It also fixes my original issue with rollup-plugin-node-resolve.

@ljharb let me know how I can help get this in. I'm not at all familiar with the preserve-symlinks-main behavior but I could learn!

**Comment by ljharb** [comment]
k, looked into https://github.com/nodejs/node/pull/19911/ and replicated its test case; it seems `preserveSymlinksMain` is only something that applies to node itself, and doesn't need any special handling in `resolve`.

More test cases would be great, but also the appveyor Windows tests seem to be failing :-/

**Comment by ljharb** [comment]
Seems like this caused https://github.com/rollup/rollup-plugin-commonjs/issues/400 :-/ hopefully we can roll forward instead of reverting

**Comment by octogonz** [comment]
😊 At least it wasn't a PATCH bump!

**Pull request merged**